### PR TITLE
Opencl blame shifting

### PR DESCRIPTION
**Pull request by aarontcopal2**
- [x] Check that Blame shift works for OpenCL
- [ ] Check that Blame shift works for Cuda
- [ ] Check that Blame shift works for Rocm

### CODE REVIEW

**Comment by Jokeren** [comment]
Can you describe the main features of this pull request? e.g., What problems you would like to target? What solutions you adopted to solve the problems? Is there any preliminary result for any application?

**Comment by aarontcopal2** [comment]
> Can you describe the main features of this pull request?

In this pull request, I have implemented blame-shifting analysis for OpenCL. This is done after referring to Milind Chabbi's paper and his old branch in hpctoolkit on the same.

> What problems you would like to target?

With this pull request, the user can enable blame shifting by passing the CPU_GPU_IDLE event. e.g. `hpcrun -e gpu=opencl -e CPU_GPU_IDLE -e REALTIME ./application`
Primarily, we can see when both the CPU and GPU are idle. 3 metrics will be visible in hpcviewer: CPU_IDLE, GPU_IDLE and CPU_IDLE_CAUSE. 

> What solutions you adopted to solve the problems? 

I have created a generic interface for blame-shifting. I have used OpenCL api wrappers at certain points that call sync_start, sync_end, kernel_start, kernel_end, queue_start and queue_end functions of the blame-shifting interface. 

Since the interface is generic, other languages/platforms can use this to enable blame-shifting for itself.

>  Is there any preliminary result for any application?

I have tried this for Pelec. But since it has a small execution time, the metric values related to blame-shifting is small. So I plan to run this with a larger example. Im trying to build amr_wind for testing this out, but not yet successful.

**Comment by mxz297** [comment]
@aarontcopal2 
 
> Since the interface is generic, other languages/platforms can use this to enable blame-shifting for itself.

I will evaluate this new interface next week to see if I can use it for AMD GPUs. 

> I have tried this for Pelec. But since it has a small execution time, the metric values related to blame-shifting is small. So I plan to run this with a larger example. Im trying to build amr_wind for testing this out, but not yet successful.

For PeleC, I thought you can increase the number of iterations at command line?

**Comment by aarontcopal2** [comment]
@mxz297 

Sure. You can try it out and let me know.

> For PeleC, I thought you can increase the number of iterations at command line?

`./PeleC3d.dpcpp.ex inputs_ex max_step=10` is the command I use to run the application. can max_step take any natural number?

**Comment by mxz297** [comment]
@aarontcopal2  I think  you can increase max_step to a larger integer to let it run longer.

**Comment by Jokeren** [comment]
> I have created a generic interface for blame-shifting.

I suppose currently only OpenCL implementation is available? It might need some extra work to support CUDA. Milind previously used cuda events to monitor GPU kernels, but now we switch to CUPTI. I'll review your code and check it out.

> the metric values related to blame-shifting is small.

You can measure relative metrics. For example, the ratio of time spent in synchronization versus the total amount of time.

File: src/tool/hpcrun/gpu/gpu-activity.h [hunk 1]
```diff
@@ -325,6 +326,13 @@ typedef struct gpu_branch_t {
 } gpu_branch_t;
 
 
+typedef struct gpu_blame_shift_t {
+  double cpu_idle_time;
+  double gpu_idle_time;
+  double cpu_idle_cause_time;
```
Comment (Jokeren) [inline src/tool/hpcrun/gpu/gpu-activity.h:332]: What is `cpu_idle_cause_time` and why there's no `gpu_idle_cause_time`?

File: src/tool/hpcrun/gpu/opencl/opencl-api.c [hunk 2]
```diff
@@ -1207,6 +1271,74 @@ hpcrun_clCreateBuffer
...
+	// clWaitForEvents can wait on multiple events(probably from different queues).
+	// We need a more complicated approach of finding the queues on which the CPU will wait
+	// For now we pass the 1st queue
+	void *queue_ptr;
+	cl_command_queue queue;
+
+	if(is_opencl_blame_shifting_enabled()) {
+		size_t queue_size;
+		clGetEventInfo(*event_list, CL_EVENT_COMMAND_QUEUE, 0, NULL, &queue_size);
+		queue_ptr = hpcrun_malloc(sizeof(queue_size));
```
Comment (Jokeren) [inline src/tool/hpcrun/gpu/opencl/opencl-api.c:1292]: Where this `queue_ptr` is recycled?

File: src/tool/hpcrun/gpu/gpu-activity.h (hunk 1 above)
Comment (aarontcopal2) [reply src/tool/hpcrun/gpu/gpu-activity.h:332]: **how is cpu_idleness detected?**
call to sync() makes the cpu thread wait. In this case we know who is waiting(cpu thread) and who is the cause for wait(either the gpu stream or a kernel inside the stream). Thus we can increment both cpu_idle time and cpu_idle_cause appropriately.

**How is gpu idleness detected?** 
itimer callbacks check the number of active kernels periodically. When (#kernels == 0), the gpu_idle_time is incremented. Note that the stacktrace where the metric is recorded is responsible for the gpu idleness. gpu_idle_cause_time does not make sense. Where would we attribute that metric? And how will it be calculated?

File: src/tool/hpcrun/sample-sources/gpu/gpu-blame/active-kernels-map.c [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+
+void
+active_kernels_delete
+(
+ uint64_t ak_id
+)
+{
+  spinlock_lock(&ak_map_lock);
```
Comment (Jokeren) [inline src/tool/hpcrun/sample-sources/gpu/gpu-blame/active-kernels-map.c:147]: If this function is called by `clear_recursive`, you will perform `lock` and `unlock` the number of times equal to the number of nodes in the map.

File: src/tool/hpcrun/sample-sources/ss-list.h [hunk 4]
```diff
@@ -104,4 +104,5 @@ SAMPLE_SOURCE_DECL_MACRO(level0)
 SAMPLE_SOURCE_DECL_MACRO(opencl)
 #endif
 #endif
+SAMPLE_SOURCE_DECL_MACRO(cpu_gpu_idle)
```
Comment (Jokeren) [inline src/tool/hpcrun/sample-sources/ss-list.h:107]: Why not rename `gpu_blame` to `cpu_gpu_idle` to keep it more consistent.

File: src/tool/hpcrun/sample-sources/gpu/gpu-blame/blame.c [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+add_kernel_to_completed_list
+(
+ kernel_node_t *kernel_node
+)
+{
+  while (true) {
+    kernel_node_t *current_head = atomic_load(&completed_kernel_list_head);
+    atomic_store(&kernel_node->next, current_head);
+    if (atomic_compare_exchange_strong(&completed_kernel_list_head, &current_head, kernel_node)) {
+      break;
```
Comment (Jokeren) [inline src/tool/hpcrun/sample-sources/gpu/gpu-blame/blame.c:202]: The above code is confusing to me. 

If you just want to append a node atomically to a concurrent list's head, you can follow the simple pattern below. No need to use an additional atomic load for list head.

https://github.com/HPCToolkit/hpctoolkit/blob/d53f6fce018673bc886305330ca91b686844468e/src/lib/prof-lean/stacks.c#L230

File: src/tool/hpcrun/gpu/gpu-activity.h (hunk 1 above)
Comment (Jokeren) [reply src/tool/hpcrun/gpu/gpu-activity.h:332]: By reviewing Milind's paper, I reminded what `cpu_idle_cause` means. In fact, the original paper denotes `gpu_idle_time` as `gpu_idle_cause_time`.

**Comment by Jokeren** [comment]
@aarontcopal2 Overall this pull request looks good, but needs very close attention to the concurrent data structures you are extensively using. I will take a second round look when I try to extend it for CUDA. 
One direction to further explore is to implement the full history blaming feature mentioned in Figure 12 of Milind's paper.

File: src/tool/hpcrun/gpu/gpu-activity.h (hunk 1 above)
Comment (aarontcopal2) [reply src/tool/hpcrun/gpu/gpu-activity.h:332]: done

File: src/tool/hpcrun/sample-sources/gpu/gpu-blame/blame.c (hunk 5 above)
Comment (aarontcopal2) [reply src/tool/hpcrun/sample-sources/gpu/gpu-blame/blame.c:202]: @Jokeren , if the CAS in the while loop evaluates to false, where is the head variable getting updated?

**Comment by aarontcopal2** [comment]
Thanks for the comments @Jokeren.

File: src/tool/hpcrun/gpu/opencl/opencl-api.c (hunk 2 above)
Comment (aarontcopal2) [reply src/tool/hpcrun/gpu/opencl/opencl-api.c:1292]: the queue_size is variable. A simple alloc and free queue won't work. how is dynamic-sized hpcrun_malloc freed in hpcrun?

File: src/tool/hpcrun/sample-sources/gpu/gpu-blame/blame.c (hunk 5 above)
Comment (Jokeren) [reply src/tool/hpcrun/sample-sources/gpu/gpu-blame/blame.c]: The head variable is updated by `atomic_compare_exchange_strong`

File: src/tool/hpcrun/gpu/opencl/opencl-api.c (hunk 2 above)
Comment (Jokeren) [reply src/tool/hpcrun/gpu/opencl/opencl-api.c]: If it is in an application callback, just use malloc.
If only the current scope uses this variable, just declare it as a stack variable.

**Pull request closed without merging**